### PR TITLE
[FLINK-18286] Implement type inference for GET/FLATTEN 

### PR DESCRIPTION
**Pull request by dawidwys**
## What is the purpose of the change

This PR implements type inference for GET/FLATTEN expressions. Additionally it aligns the behaviour between Table API & SQL. To achieve that it fixes handling nullability of nested columns in operators such as DOT, ITEM and nested fields coming directly from a Table.

## Verifying this change

Added tests in 
* `InputTypeStrategiesTest` for `COMPOSITE` strategy
* `TypeStrategiesTest` for `GET` strategy
* `CompositeTypeAccessExpressionITCase` IT tests for accessing nested fields in both SQL & Table API

## Does this pull request potentially affect one of the following parts:

  - Dependencies (does it add or upgrade a dependency): (yes / **no**)
  - The public API, i.e., is any changed class annotated with `@Public(Evolving)`: (yes / **no**)
  - The serializers: (yes / **no** / don't know)
  - The runtime per-record code paths (performance sensitive): (yes / **no** / don't know)
  - Anything that affects deployment or recovery: JobManager (and its components), Checkpointing, Kubernetes/Yarn/Mesos, ZooKeeper: (yes / **no** / don't know)
  - The S3 file system connector: (yes / **no** / don't know)

## Documentation

  - Does this pull request introduce a new feature? (yes / **no**)
  - If yes, how is the feature documented? (**not applicable** / docs / JavaDocs / not documented)

### CODE REVIEW

**Comment by dawidwys** [comment]
@danny0405 Could you check changes in 11f71d1b13e2b6456722b5ce5f1081a6f9d1cef9 in classes:
* `SqlDotOperator`
* `SqlItemOperator`
* `SqlValidatorScope`

and say if they make sense and if we can also get those changes into Calcite? (I think those are fixes for valid bugs in Calcite).

**Comment by dawidwys** [comment]
@twalthr Could you review the PR?

**Comment by dawidwys** [comment]
Let me explain what the changes are about. They are not trying to fix the nullability of the record type in all operators, but in my opinion it actually improves the behaviour described in the thread you posted (thank you for it, it is actually a good read).

First of all we do very much want to respect the original nullability, that's the main purpose of those changes and also why we have a special handling in `FlinkTypeFactory:348`. We have that because Calcite actually modifies nested columns of a `ROW` type, if the type is nullable.

In calcite a type `ROW<i INT NOT NULL>` becomes `ROW<i INT>` See `org.apache.calcite.rel.type.RelDataTypeFactoryImpl#createTypeWithNullability:326-338`. This behaviour is especially problematic for STRUCTURED types, which correspond to java's pojos. Let's take a look at an example:
```
class Address {
  int id;
  String city;
}
```
such a pojo we would like to map to a `Address(id INT NOT NULL, city STRING)`, where the `Address` type itself (outer row) is nullable, but the nested `id` field is `NOT NULL`. This is not possible in Calcite by default. Thats why we have the changes in `org.apache.calcite.rel.type.RelDataTypeFactoryImpl#createTypeWithNullability:326-338`.

On the other hand if we access a nested field of a nullable composite type we do want to adjust the nullability, cause accessing a field of a null row should obviously produce null values. If we do
```
CREATE TABLE test (
  address Address
);

// the type should be nullable INT, if the outer row is null we can not produce a not null id
SELECT t.address.id FROM test t;
```

BTW this behaviour is already present in `SqlDotOperator#inferReturnType`, but as far as I can tell that method is not used.

To summarize it again. I do very much want to respect the nullability of the record attributes. Thats the main goal of the changes. Calcite as of now does not respect the nullability of nested fields, which we had to fix in FLINK-16344. The changes here affect the return type of field accessors, not the record itself.

I hope it gives a better overview what are the changes. If you have any questions let me know, I will try to elaborate more.

**Comment by danny0405** [comment]
> This is not possible in Calcite by default. Thats why we have the changes in `org.apache.calcite.rel.type.RelDataTypeFactoryImpl#createTypeWithNullability:326-338`.

The calcite code says:

> If nullable -> Do a deep copy, setting all fields of the record type
> to be nullable regardless of initial nullability.

This seems just what you need, what do you mean it is not possible ?

**Comment by dawidwys** [comment]
Yes, but does it always. We want that behaviour only when accessing the field.

Imagine a user function that accepts `ADDRESS(id INT NOT NULL, city STRING)`.

```
class UDF extends ScalarFunction {
  public Object eval(Address address) {
     if (address != null) {
         // based on the type I know address.id is NOT NULL here
        int id = address.id;
     }
  }
}
```

but this is not possible in Calcite. There is no way to construct a `ADDRESS(id INT NOT NULL, city STRING)`. It will be changed to `ADDRESS(id INT, city STRING)`. And based on that there is no way to have the behaviour as above. The id is always nullable.



On the other hand I want the type adjusted if I pass the field only, as this might be null without additional checks:

```
class UDF2 extends ScalarFunction {
  public Object eval(int id {
  }
}

// This is illegal. We can not ensure that t.address.id is NOT NULL
SELECT UDF2(t.address.id) FROM test t;
```

**Comment by danny0405** [comment]
Got you ~ The conclusion in the JIRA issue is synced with your needs: the nullability of structure type should not affect its members. I would try to fire a fix in Calcite, there may needs more discussion there ~

File: flink-table/flink-table-planner-blink/src/main/scala/org/apache/flink/table/planner/codegen/ExprCodeGenerator.scala [hunk 1]
```diff
@@ -724,6 +724,9 @@ class ExprCodeGenerator(ctx: CodeGeneratorContext, nullableInput: Boolean)
             val key = operands(1)
             generateMapGet(ctx, operands.head, key)
 
+          case t: LogicalType if TypeCheckUtils.isRow(t) =>
```
Comment (twalthr) [inline flink-table/flink-table-planner-blink/src/main/scala/org/apache/flink/table/planner/codegen/ExprCodeGenerator.scala:727]: use `LogicalTypeChecks` instead

File: flink-table/flink-table-planner-blink/src/test/java/org/apache/flink/table/planner/expressions/CompositeTypeAccessExpressionITCase.java [hunk 2]
```diff
@@ -0,0 +1,264 @@
...
+
+				// Actually in case of SQL it does not use the GET method, but
+				// a custom logic for accessing nested fields of a Table.
+				TestSpec.forFunction(BuiltInFunctionDefinitions.GET)
+					.onFieldsWithData(null, Row.of(1))
+					.andDataTypes(
+						ROW(FIELD("nested", BIGINT().notNull())).nullable(),
+						ROW(FIELD("nested", BIGINT().notNull())).notNull()
+					)
+					.testTableApiResult($("f0").get("nested"), null, BIGINT().nullable())
```
Comment (twalthr) [inline flink-table/flink-table-planner-blink/src/test/java/org/apache/flink/table/planner/expressions/CompositeTypeAccessExpressionITCase.java:88]: use the newly introduced `testResult` such that a result must be defined only once

File: flink-table/flink-table-planner-blink/src/test/java/org/apache/flink/table/planner/expressions/CompositeTypeAccessExpressionITCase.java [hunk 3]
```diff
@@ -0,0 +1,264 @@
...
+						ARRAY(BIGINT().notNull()).notNull(),
+						MAP(STRING(), BIGINT().notNull()).nullable(),
+						MAP(STRING(), BIGINT().notNull()).notNull(),
+						ROW(FIELD("nested", BIGINT().notNull())).nullable(),
+						ROW(FIELD("nested", BIGINT().notNull())).notNull()
+					)
+					// accessing elements of MAP or ARRAY is a runtime operations,
+					// we do not know about the size or contents during the inference
+					// therefore the results are always nullable
+					.testSqlResult("f0[1]", null, BIGINT().nullable())
```
Comment (twalthr) [inline flink-table/flink-table-planner-blink/src/test/java/org/apache/flink/table/planner/expressions/CompositeTypeAccessExpressionITCase.java:107]: should we port `AT` as well or in a separate PR?

File: flink-table/flink-table-planner-blink/src/test/java/org/apache/flink/table/planner/expressions/CompositeTypeAccessExpressionITCase.java [hunk 4]
```diff
@@ -0,0 +1,264 @@
...
+					.testSqlResult("f4['nested']", null, BIGINT().nullable())
+					.testSqlResult("f5['nested']", 1L, BIGINT().notNull())
+			);
+		}
+	}
+
+	/**
+	 * A class for customized tests.
+	 */
+	public static class CallFieldAccess {
```
Comment (twalthr) [inline flink-table/flink-table-planner-blink/src/test/java/org/apache/flink/table/planner/expressions/CompositeTypeAccessExpressionITCase.java:123]: nit: very good tests in general but can we rename the two suits to `FieldAccessAfterCall` and `FieldAccessFromTable` or similar to highlight the difference?

File: flink-table/flink-table-planner-blink/src/test/java/org/apache/flink/table/planner/expressions/CompositeTypeAccessExpressionITCase.java (hunk 3 above)
Comment (dawidwys) [reply flink-table/flink-table-planner-blink/src/test/java/org/apache/flink/table/planner/expressions/CompositeTypeAccessExpressionITCase.java:107]: I have it in a different branch. I found this PR already quite big.

File: flink-table/flink-table-common/src/main/java/org/apache/flink/table/types/inference/TypeStrategies.java [hunk 5]
```diff
@@ -315,6 +317,36 @@ public static TypeStrategy nullable(TypeStrategy initialStrategy) {
...
 	};
 
+	/**
+	 * Type strategy that returns a type of a field nested inside a composite type that is described by the second argument.
+	 * The second argument must be a literal that describes either the nested field name or index.
+	 */
+	public static final TypeStrategy GET = callContext -> {
+		List<DataType> argumentDataTypes = callContext.getArgumentDataTypes();
+		DataType rowDataType = argumentDataTypes.get(0);
+		TableSchema nestedSchema = DataTypeUtils.expandCompositeTypeToSchema(rowDataType);
```
[the remark this answers is not in the document]
Comment (dawidwys) [reply flink-table/flink-table-common/src/main/java/org/apache/flink/table/types/inference/TypeStrategies.java:327]: I think it does already support the `DISTINCT` type. Moreover `AtomicDataType` can not hold a composite type. `AtomicDataType` can not hold the bridging classes of nested fields. IMO `Distinct` type can be any of `FieldsDataType`, `AtomicDataType`, 'KeyValueDataType` etc.

The way we support `DISTINCT` type is that we do not check the `LogicalTypeRoot` in `expandCompositeTypeToSchema`, but we check only for the `FieldsDataType`.

File: flink-table/flink-table-planner-blink/src/main/scala/org/apache/flink/table/planner/codegen/ExprCodeGenerator.scala [hunk 6]
```diff
@@ -719,18 +721,18 @@ class ExprCodeGenerator(ctx: CodeGeneratorContext, nullableInput: Boolean)
...
             requireInteger(index)
             generateArrayElementAt(ctx, array, index)
 
-          case t: LogicalType if TypeCheckUtils.isMap(t) =>
+          case LogicalTypeRoot.MAP =>
             val key = operands(1)
             generateMapGet(ctx, operands.head, key)
 
-          case t: LogicalType if TypeCheckUtils.isRow(t) =>
+          case LogicalTypeRoot.ROW =>
```
Comment (twalthr) [inline flink-table/flink-table-planner-blink/src/main/scala/org/apache/flink/table/planner/codegen/ExprCodeGenerator.scala:735]: LogicalTypeRoot.ROW | LogicalTypeRoot.STRUCTURED_TYPE

File: flink-table/flink-table-planner-blink/src/main/java/org/apache/calcite/sql/fun/SqlDotOperator.java [hunk 7]
```diff
@@ -0,0 +1,194 @@
...
+    final RelDataTypeField field =
+        nodeType.getField(fieldName, false, false);
+    if (field == null) {
+      throw SqlUtil.newContextException(fieldId.getParserPosition(),
+          Static.RESOURCE.unknownField(fieldName));
+    }
+    RelDataType type = field.getType();
+    if (nodeType.isNullable() && !type.isNullable()) {
+        type = validator.getTypeFactory().createTypeWithNullability(type, true);
+    }
```
Comment (danny0405) [inline flink-table/flink-table-planner-blink/src/main/java/org/apache/calcite/sql/fun/SqlDotOperator.java:119]: In Calcite, when the record type is nullable, the attributes are always nullable, so, this fix is not that necessary from the Calcite side.

**Comment by dawidwys** [comment]
Some of the changes were contributed to Calcite (changes that made us copy some of Calcite classes). See: https://github.com/apache/calcite/commit/e2942feef4d52c577a46cc3f2b92fa462035bb20

Therefore I will merge it once I get a green build. Unless there are objections.
